builder/dockerfile/builder.go

```diff
@@ -147,23 +147,22 @@ func (b *Builder) build(source builder.Source, dockerfile *parser.Result) (*buil
 
 	addNodesForLabelOption(dockerfile.AST, b.options.Labels)
 
-	if err := checkDispatchDockerfile(dockerfile.AST); err != nil {
-		buildsFailed.WithValues(metricsDockerfileSyntaxError).Inc()
-		return nil, err
-	}
-
-	dispatchState, err := b.dispatchDockerfileWithCancellation(dockerfile, source)
+	stages, metaArgs, err := typedcommand.Parse(dockerfile.AST, b.options.Target, buildsFailed)
 	if err != nil {
 		return nil, err
 	}
 
-	if b.options.Target != "" && !dispatchState.isCurrentStage(b.options.Target) {
+	if b.options.Target != "" && !stages.IsCurrentStage(b.options.Target) {
```

Comment:
I agree, only problem is variable expansion. Is it ok if we don't support dockerfiles with meta args like:
```
ARG FOO=BAR
ARG BAZ=${FOO}:FOO
FROM <...>
```
?